implementation/clabbert/src/main/java/pwr/jakprzyjade/clabbert/api/middlewares/UserHeaderInterceptor.java

```diff
@@ -0,0 +1,73 @@
+/* @jakubzehner (C) 2024 */
+package pwr.jakprzyjade.clabbert.api.middlewares;
+
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import org.springframework.lang.NonNull;
+import org.springframework.stereotype.Component;
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.HandlerInterceptor;
+import pwr.jakprzyjade.clabbert.api.annotations.UserRoles;
+import pwr.jakprzyjade.clabbert.domain.UserRole;
+import pwr.jakprzyjade.clabbert.domain.exceptions.authorization.UserIdHeaderMissing;
+import pwr.jakprzyjade.clabbert.domain.exceptions.authorization.UserRoleHeaderMissing;
+import pwr.jakprzyjade.clabbert.domain.exceptions.authorization.UserRoleNotSupported;
+import pwr.jakprzyjade.clabbert.domain.exceptions.authorization.UserUnauthorized;
+
+@Component
+public class UserHeaderInterceptor implements HandlerInterceptor {
+
+    private static final List<String> supportedRoles =
+            Arrays.stream(UserRole.values()).map(Enum::name).toList();
+
+    @Override
+    public boolean preHandle(
+            @NonNull HttpServletRequest request,
+            @NonNull HttpServletResponse response,
+            @NonNull Object handler)
+            throws Exception {
+        var userId = request.getHeader("user-id");
+        var userRole = request.getHeader("user-role");
+
+        if (userRole == null) {
+            throw new UserRoleHeaderMissing();
+        }
+
+        if (!supportedRoles.contains(userRole.toUpperCase())) {
+            throw new UserRoleNotSupported();
+        }
+
+        var userRoleEnum = UserRole.valueOf(userRole.toUpperCase());
+        var requiredRoles = getRequiredRoles(handler);
+
+        if (userId == null && userRoleEnum != UserRole.GUEST) {
+            throw new UserIdHeaderMissing();
+        }
+
+        if (requiredRoles.isPresent() && !List.of(requiredRoles.get()).contains(userRoleEnum)) {
+            throw new UserUnauthorized();
+        }
+
+        request.setAttribute("userId", userId);
+        request.setAttribute("userRole", userRoleEnum);
+        return true;
+    }
+
+    private Optional<UserRole[]> getRequiredRoles(Object handler) {
```

Comment:
**nit:** Probably needs `@NonNull` somewhere, since `Optional` can be `empty`, `of`, but also `null` (I love Java). Looking now, wouldn't most of your methods require `@NonNull` return type annotation?

Reply:
On the one hand this is true, on the other hand it seems to me that the minimum responsibility should be on the developer's side and if the Optional type is used then I assume in advance that there will never be a situation where null is returned.
Especially since even the description of this class in Java has such a piece of description:
> API Note:
>Optional is primarily intended for use as a method return type where there is a clear need to represent "no result," and where using null is likely to cause errors. A variable whose type is Optional should never itself be null; it should always point to an Optional instance.

In addition, I have no intention of placing the `@NonNull` annotation everywhere until the static analyzer tells me that it is expected in a given place.